test/infrastructure/container/docker.go

```diff
@@ -0,0 +1,400 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package container provides an interface for interacting with Docker and potentially
+// other container runtimes.
+package container
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+var dockerClient *client.Client
+
+// getDockerClient returns a new client connection for interacting with the Docker engine.
+func getDockerClient() (*client.Client, error) {
+	if dockerClient == nil {
+		var err error
+		dockerClient, err = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+		if err != nil {
+			return nil, fmt.Errorf("create Docker client: %v", err)
+		}
+	}
+	return dockerClient, nil
+}
+
+// SaveContainerImage saves a Docker image to the file specified by dest.
+func SaveContainerImage(ctx context.Context, image, dest string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	reader, err := docker.ImageSave(ctx, []string{image})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tar, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer tar.Close()
+
+	_, err = io.Copy(tar, reader)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PullContainerImage triggers the Docker engine to pull an image if it is not
+// already present.
+func PullContainerImage(ctx context.Context, image string) error {
```

Comment:
```suggestion
func pullContainerImage(ctx context.Context, image string) error {
```
This seems used only in this module

Reply:
Right now (with this change) this is only called internally to this module. But I think we will want the ability for consuming code to just pull images. Is there a reason you would like to make this private?

Reply:
I you expect to use it in CAPD ok; I was just trying to reduce the surface of this package to the minimum required

Reply:
Makes sense. I will make sure to check at the end of this work that that's really the case. If it turns out there is no need for it, I'll be sure to propose an update to switch it to a private function. Thanks!